core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -161,30 +172,39 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
    * Update the partition states of multiple partitions in zookeeper.
    * @param leaderAndIsrs The partition states to update.
    * @param controllerEpoch The current controller epoch.
+   * @param controllerEpochVersion expected controller epoch zkVersion to check.
    * @return UpdateLeaderAndIsrResult instance containing per partition results.
    */
-  def updateLeaderAndIsr(leaderAndIsrs: Map[TopicPartition, LeaderAndIsr], controllerEpoch: Int): UpdateLeaderAndIsrResult = {
+  def updateLeaderAndIsr(leaderAndIsrs: Map[TopicPartition, LeaderAndIsr], controllerEpoch: Int, controllerEpochVersion: Int): UpdateLeaderAndIsrResult = {
     val successfulUpdates = mutable.Map.empty[TopicPartition, LeaderAndIsr]
     val updatesToRetry = mutable.Buffer.empty[TopicPartition]
     val failed = mutable.Map.empty[TopicPartition, Exception]
     val leaderIsrAndControllerEpochs = leaderAndIsrs.map { case (partition, leaderAndIsr) =>
       partition -> LeaderIsrAndControllerEpoch(leaderAndIsr, controllerEpoch)
     }
-    val setDataResponses = try {
-      setTopicPartitionStatesRaw(leaderIsrAndControllerEpochs)
+    val multiOpResponses = try {
+      setTopicPartitionStatesRaw(leaderIsrAndControllerEpochs, controllerEpochVersion)
     } catch {
       case e: Exception =>
         leaderAndIsrs.keys.foreach(partition => failed.put(partition, e))
         return UpdateLeaderAndIsrResult(successfulUpdates.toMap, updatesToRetry, failed.toMap)
     }
-    setDataResponses.foreach { setDataResponse =>
-      val partition = setDataResponse.ctx.get.asInstanceOf[TopicPartition]
-      setDataResponse.resultCode match {
+    multiOpResponses.foreach { multiOpResponse =>
+      val partition = multiOpResponse.ctx.get.asInstanceOf[TopicPartition]
+      multiOpResponse.resultCode match {
         case Code.OK =>
-          val updatedLeaderAndIsr = leaderAndIsrs(partition).withZkVersion(setDataResponse.stat.getVersion)
+          val setDataResult = multiOpResponse.opResults.get(1).asInstanceOf[SetDataResult]
+          val updatedLeaderAndIsr = leaderAndIsrs(partition).withZkVersion(setDataResult.getStat.getVersion)
           successfulUpdates.put(partition, updatedLeaderAndIsr)
-        case Code.BADVERSION => updatesToRetry += partition
-        case _ => failed.put(partition, setDataResponse.resultException.get)
+        case Code.BADVERSION => {
```

Comment:
nits: we probably don't need `{` here so that the code style is consistent with the existing code.

Reply:
Thanks for pointing out. Will fix that.